### PR TITLE
fix: :bug: Hotfix. Change selectors due to new html structure

### PR DESCRIPTION
**Pull request by chakaponi**
LinkedIn has changed their HTML structure therefore app stopped to apply to jobs

### CODE REVIEW

**Comment by chakaponi** [comment]
Sorry, I supposed to begin branch name with "hotfix" instead of "bug"

**Comment by chakaponi** [comment]
Currently, only the selectors that I have changed have been tested for changes on LinkedIn side. There might be more

**Comment by surapuramakhil** [comment]
@chakaponi released hotfix, I just applied - automation didn't break. can you explain on impact

**Comment by chakaponi** [comment]
@surapuramakhil Are we ready to deliver fix?

**Comment by chakaponi** [comment]
Going to fix #919

**Comment by surapuramakhil** [comment]
@chakaponi can you create your branch from main? so that we can do hotfix.

**Comment by chakaponi** [comment]
@surapuramakhil Ready to be delivered. All tests passed
After merge you may close issues #920, #919 and the current one

**Comment by surapuramakhil** [comment]
> @surapuramakhil Ready to be delivered. All tests passed After merge you may close issues #920, #919 and the current one

tested just now, didn't work

**Comment by WayneWu01** [comment]
it shows the same results as before no job found

**Comment by chakaponi** [comment]
@WayneWu01 @surapuramakhil You may share jobs container's HTML to my email so I can check differences and get some ideas regarding the way LinkedIn change their structure
If not, someone would be have to get XPATH by sending html to AI-model as well

**Comment by surapuramakhil** [comment]
@chakaponi can we connect on discord? so that things can be done faster.

**Comment by chakaponi** [comment]
@feder-cr We need help with undestanding 'pb4' css-class destiny. It doesn't exist anymore. Code hadn't been documented so we lost context of this class application area.
[Source](https://github.com/AIHawk-FOSS/Auto_Jobs_Applier_AI_Agent/blob/9455bfa85a0e003c9afcb3300222eedf2ad05b4b/src/ai_hawk/linkedIn_easy_applier.py#L379)

**Comment by chakaponi** [comment]
@surapuramakhil Discord's having auth problems so I can't use it. Only [telegram](t.me/chakaponi)

**Comment by surapuramakhil** [comment]
@chakaponi not able to find you. can you ping me on telegram

**Comment by surapuramakhil** [comment]
thread link for sharing your jobs view structure, for having a fix which works for all.

https://discord.com/channels/1300208460788400159/1310293395226230896

run this command on dev tools console 

> copy($('#main'))

you can run copy($('#main')) in your chrome dev tools on, for example, this page: https://www.linkedin.com/jobs/search/?currentJobId=4042558197&distance=10&f_E=4&f_LF=f_AL&f_WT=2&keywords=JavaScript%20Developer&location=Mexico  It will copy element which contain currently open job and block with list of jobs (left sidebar)

**Comment by feder-cr** [comment]
@chakaponi hello, if it is no longer useful remove it with no problem!

**Comment by chakaponi** [comment]
@surapuramakhil Hotfix branch has been merged into mine with some changes and rewritten tests. All tests pass. Works in normal environment for me.

File: src/ai_hawk/job_manager.py [hunk 1]
```diff
@@ -254,23 +253,22 @@ def get_jobs_from_page(self, scroll=False):
             pass
 
         try:
-            # XPath query to find the ul tag with class scaffold-layout__list-container
-            job_results_xpath_query = "//ul[contains(@class, 'scaffold-layout__list-container')]"
-            job_results = self.driver.find_element(By.XPATH, job_results_xpath_query)
+            jobs_container = self.driver.find_element(By.CLASS_NAME, 'scaffold-layout__list-container')
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:256]: revert this change

File: src/ai_hawk/job_manager.py [hunk 2]
```diff
@@ -254,23 +253,22 @@ def get_jobs_from_page(self, scroll=False):
...
-                logger.warning(f'is scrollable: {browser_utils.is_scrollable(job_results_scrolableElament)}')
+                jobs_container_scrolableElement = jobs_container.find_element(By.XPATH,"..")
+                logger.warning(f'is scrollable: {browser_utils.is_scrollable(jobs_container_scrolableElement)}')
+
+                browser_utils.scroll_slow(self.driver, jobs_container_scrolableElement)
+                browser_utils.scroll_slow(self.driver, jobs_container_scrolableElement, step=300, reverse=True)
 
-                browser_utils.scroll_slow(self.driver, job_results_scrolableElament)
-                browser_utils.scroll_slow(self.driver, job_results_scrolableElament, step=300, reverse=True)
+            job_element_list = jobs_container.find_elements(By.CSS_SELECTOR, 'div[data-job-id]')
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:265]: rever this

File: src/ai_hawk/job_manager.py [hunk 3]
```diff
@@ -254,23 +253,22 @@ def get_jobs_from_page(self, scroll=False):
...
         try:
-            # XPath query to find the ul tag with class scaffold-layout__list-container
-            job_results_xpath_query = "//ul[contains(@class, 'scaffold-layout__list-container')]"
-            job_results = self.driver.find_element(By.XPATH, job_results_xpath_query)
+            jobs_container = self.driver.find_element(By.CLASS_NAME, 'scaffold-layout__list-container')
 
             if scroll:
-                job_results_scrolableElament = job_results.find_element(By.XPATH,"..")
-                logger.warning(f'is scrollable: {browser_utils.is_scrollable(job_results_scrolableElament)}')
+                jobs_container_scrolableElement = jobs_container.find_element(By.XPATH,"..")
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:259]: defend your variable name - why jobs_container_scrolableElement is more correct naming than job_results_scrolableElament

File: src/ai_hawk/job_manager.py [hunk 4]
```diff
@@ -288,13 +286,14 @@ def read_jobs(self):
...
             pass
         
-        job_results = self.driver.find_element(By.CLASS_NAME, "jobs-search-results-list")
-        browser_utils.scroll_slow(self.driver, job_results)
-        browser_utils.scroll_slow(self.driver, job_results, step=300, reverse=True)
-        job_list_elements = self.driver.find_elements(By.CLASS_NAME, 'scaffold-layout__list-container')[0].find_elements(By.CLASS_NAME, 'jobs-search-results__list-item')
-        if not job_list_elements:
-            raise Exception("No job class elements found on page")
-        job_list = [self.job_tile_to_job(job_element) for job_element in job_list_elements] 
+        jobs_container = self.driver.find_element(By.CLASS_NAME, 'scaffold-layout__list-container')
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:289]: refactor this - "DRY"

File: src/ai_hawk/job_manager.py [hunk 5]
```diff
@@ -494,9 +481,9 @@ def job_tile_to_job(self, job_tile) -> Job:
             logger.debug(f"Job link extracted: {job.link}")
         except NoSuchElementException:
             logger.warning("Job link is missing.")
-        
+
         try:
-            job.company = job_tile.find_element(By.XPATH, ".//div[contains(@class, 'artdeco-entity-lockup__subtitle')]//span").text
+            job.company = job_tile.find_element(By.XPATH, './/span[contains(normalize-space(), " · ")]').text.split(' · ')[0].strip()
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:486]: there could be many spans. revert this

File: src/ai_hawk/job_manager.py [hunk 6]
```diff
@@ -513,12 +500,12 @@ def job_tile_to_job(self, job_tile) -> Job:
             logger.warning(f"Failed to extract job ID: {e}", exc_info=True)
 
         try:
-            job.location = job_tile.find_element(By.CLASS_NAME, 'job-card-container__metadata-item').text
+            job.location = job_tile.find_element(By.XPATH, './/span[contains(normalize-space(), " · ")]').text.split(' · ')[-1].strip()
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:503]: this do works why mess with it?

Explain your xpath, how are you making sure that span is location? are there any strong identifier this will for sure will be location.

File: src/ai_hawk/job_manager.py [hunk 7]
```diff
@@ -513,12 +500,12 @@ def job_tile_to_job(self, job_tile) -> Job:
...
         except NoSuchElementException:
             logger.warning("Job location is missing.")
         
         try:
-            job.apply_method = job_tile.find_element(By.CLASS_NAME, 'job-card-container__apply-method').text
+            job.apply_method = job_tile.find_element(By.XPATH, ".//div[contains(@class, 'job-card-container__job-insight-text') and normalize-space() = 'Easy Apply']").text
```
Comment (surapuramakhil) [inline src/ai_hawk/job_manager.py:508]: correct this. Also include this bug fix https://github.com/AIHawk-FOSS/Auto_Jobs_Applier_AI_Agent/issues/927

remove equals "Easy Apply", consider state where Applied, or saved previous application, as of now bot does that when it fails to submit.

File: tests/test_aihawk_job_manager.py [hunk 8]
```diff
@@ -71,21 +71,33 @@ def test_get_jobs_from_page_no_jobs(mocker, job_manager):
 
 def test_get_jobs_from_page_with_jobs(mocker, job_manager):
     """Test get_jobs_from_page when job elements are found."""
-    # Mock the no_jobs_element to behave correctly
-    mock_no_jobs_element = mocker.Mock()
-    mock_no_jobs_element.text = "No matching jobs found"
+    # Mock no_jobs_element to simulate the absence of "No matching jobs found" banner
+    no_jobs_element = mocker.Mock()
```
Comment (surapuramakhil) [inline tests/test_aihawk_job_manager.py:75]: add mock prefix or suffix.

**Review by surapuramakhil — CHANGES_REQUESTED** [review]

**Comment by surapuramakhil** [comment]
@chakaponi can you comment on https://github.com/AIHawk-FOSS/Auto_Jobs_Applier_AI_Agent/issues/927 https://github.com/AIHawk-FOSS/Auto_Jobs_Applier_AI_Agent/issues/926 so that I can assign those to you.

**Comment by Shuo-Wang-UCBerkeley** [comment]
Hi guys, I git clone the [hotfix/v2024.11.23](https://github.com/AIHawk-FOSS/Auto_Jobs_Applier_AI_Agent/tree/hotfix/v2024.11.23), but still does not work.

image

File: src/ai_hawk/job_manager.py (hunk 7 above)
Comment (surapuramakhil) [reply src/ai_hawk/job_manager.py]: done. Not complete fix of 927. but at least it marks and go

File: tests/test_aihawk_job_manager.py (hunk 8 above)
Comment (surapuramakhil) [reply tests/test_aihawk_job_manager.py]: done

**Review by surapuramakhil — APPROVED** [review]

**Comment by surapuramakhil** [comment]
@chakaponi Next time don't force push. It corrupts subsequent processes.

**Comment by Shuo-Wang-UCBerkeley** [comment]
Thank you for your effort @surapuramakhil , may I ask if the issue has fixed?
Do I need to get clone again for the AIHawk-FOSS:hotfix/v2024.11.23?

> @chakaponi Next time don't force push. It corrupts subsequent processes.

**Comment by surapuramakhil** [comment]
> Thank you for your effort @surapuramakhil , may I ask if the issue has fixed?
Do I need to get clone again for the AIHawk-FOSS:hotfix/v2024.11.23?

just update your main.
``sh
git fetch 
git pull
``

if you face errors then clone again.

File: src/ai_hawk/job_manager.py [hunk 9]
```diff
@@ -253,7 +253,9 @@ def get_jobs_from_page(self, scroll=False):
             pass
 
         try:
-            jobs_container = self.driver.find_element(By.CLASS_NAME, 'scaffold-layout__list-container')
+            # XPath query to find the ul tag with class scaffold-layout__list-container
+            jobs_xpath_query = "//ul[contains(@class, 'scaffold-layout__list-container')]"
+            jobs_container = self.driver.find_element(By.XPATH, jobs_xpath_query)
```
Comment (chakaponi) [inline src/ai_hawk/job_manager.py:258]: They unlikely will keep this class with not ul tag so both approaches are equal

File: src/ai_hawk/job_manager.py [hunk 10]
```diff
@@ -262,7 +264,7 @@ def get_jobs_from_page(self, scroll=False):
                 browser_utils.scroll_slow(self.driver, jobs_container_scrolableElement)
                 browser_utils.scroll_slow(self.driver, jobs_container_scrolableElement, step=300, reverse=True)
 
-            job_element_list = jobs_container.find_elements(By.CSS_SELECTOR, 'div[data-job-id]')
+            job_element_list = jobs_container.find_elements(By.XPATH, ".//li[contains(@class, 'jobs-search-results__list-item') and contains(@class, 'ember-view')]")
```
Comment (chakaponi) [inline src/ai_hawk/job_manager.py:267]: I would say that rely on "ember-view" instead of "data-job-id" more likely to cause problems again so I would revert this change

File: src/ai_hawk/job_manager.py [hunk 11]
```diff
@@ -483,7 +473,7 @@ def job_tile_to_job(self, job_tile) -> Job:
             logger.warning("Job link is missing.")
 
         try:
-            job.company = job_tile.find_element(By.XPATH, './/span[contains(normalize-space(), " · ")]').text.split(' · ')[0].strip()
+            job.company = job_tile.find_element(By.XPATH, ".//div[contains(@class, 'artdeco-entity-lockup__subtitle')]//span").text
```
Comment (chakaponi) [inline src/ai_hawk/job_manager.py:476]: Not robust. Also advise to revert

File: src/ai_hawk/job_manager.py [hunk 12]
```diff
@@ -500,15 +490,21 @@ def job_tile_to_job(self, job_tile) -> Job:
             logger.warning(f"Failed to extract job ID: {e}", exc_info=True)
 
         try:
-            job.location = job_tile.find_element(By.XPATH, './/span[contains(normalize-space(), " · ")]').text.split(' · ')[-1].strip()
+            job.location = job_tile.find_element(By.CLASS_NAME, 'job-card-container__metadata-item').text
```
Comment (chakaponi) [inline src/ai_hawk/job_manager.py:493]: Simply doesn't work. If we check, 'job-card-container__metadata-item' never follow as to location, but to salary fork. Has to be reverted with no doubt

File: src/ai_hawk/job_manager.py [hunk 13]
```diff
@@ -500,15 +490,21 @@ def job_tile_to_job(self, job_tile) -> Job:
...
         except NoSuchElementException:
             logger.warning("Job location is missing.")
         
+
         try:
-            job.apply_method = job_tile.find_element(By.XPATH, ".//div[contains(@class, 'job-card-container__job-insight-text') and normalize-space() = 'Easy Apply']").text
+            job_state = job_tile.find_element(By.XPATH, ".//ul[contains(@class, 'job-card-list__footer-wrapper')]//li[contains(@class, 'job-card-container__apply-method')]").text
```
Comment (chakaponi) [inline src/ai_hawk/job_manager.py:499]: Doesn't work. Return empty array all of the time. Revert

**Review by chakaponi — COMMENTED** [review]
Some bad approaches is choosen. Needs partial revert

**Comment by chakaponi** [comment]
@surapuramakhil If you want to keep your approach, you can rewrite code a little and keep both so if one class returns an empty array we could try other one right after

**Comment by surapuramakhil** [comment]
My Apologies @chakaponi as this was breaking change, I took decision for "quicker release, for pushing fix which at least work, than no fix at all".
1. we can update this in release branch (more robust)
2. only if main breaks - we can push this two second hotfix

**Comment by Shuo-Wang-UCBerkeley** [comment]
Thank you for your great work on this. @surapuramakhil and @chakaponi 
May I ask which branch works now?
so I can git clone and start the job application.

**Comment by chakaponi** [comment]
> My Apologies @chakaponi as this was breaking change, I took decision for "quicker release, for pushing fix which at least work, than no fix at all".
> 
> 1. we can update this in release branch (more robust)
> 2. only if main breaks - we can push this two second hotfix

Don't get it. My fix was ready and tested so why wouldn't have used it
Anyways, we need second hotfix as long as current fix doesn't work

**Comment by chakaponi** [comment]
> Thank you for your great work on this. @surapuramakhil and @chakaponi May I ask which branch works now? so I can git clone and start the job application.

Main branch may work if you are lucky enough. If not, try this branch, but with reverted commit with hash 22f2c3b28c16e8b5ee23931dd753015b4a1e4549

**Comment by surapuramakhil** [comment]
@chakaponi as you wished created hotfix branch. kindly raise PR to hotfix branch
